File: discovery/models/url.py
"""URL discovery data models for deep crawling with Playwright"""
from pydantic import BaseModel, Field
from typing import List, Dict, Any, Optional, Set
from datetime import datetime


class DiscoveredURL(BaseModel):
    """Individual discovered URL from deep crawling"""
    url: str = Field(..., description="Full URL")
    method: str = Field(default="GET", description="HTTP method (GET, POST, etc.)")
    source_page: Optional[str] = Field(default=None, description="Page where URL was found")
    depth: int = Field(default=0, description="Crawl depth from start")
    parameters: Dict[str, Any] = Field(default_factory=dict, description="Query params or form fields")
    discovered_at: datetime = Field(default_factory=datetime.utcnow)
    response_code: Optional[int] = Field(default=None, description="HTTP response code")
    content_type: Optional[str] = Field(default=None, description="Content-Type header")
    discovered_via: str = Field(default="deep_crawl", description="Discovery method: deep_crawl|form|js_analysis")

    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


class FormData(BaseModel):
    """Form metadata discovered during crawling"""
    action_url: str = Field(..., description="Form action URL")
    method: str = Field(default="GET", description="Form submission method")
    fields: List[Dict[str, Any]] = Field(default_factory=list, description="Form fields with metadata")
    source_page: Optional[str] = Field(default=None, description="Page containing the form")
    requires_auth: Optional[bool] = Field(default=None, description="Whether form requires authentication")

    class Config:
        arbitrary_types_allowed = True


class URLDiscoveryResult(BaseModel):
    """Results from URL discovery crawl using Playwright"""
    urls: List[DiscoveredURL] = Field(default_factory=list, description="All discovered URLs")
    forms: List[FormData] = Field(default_factory=list, description="Discovered forms")
    total_urls: int = Field(default=0, description="Total URLs discovered")
    unique_urls: int = Field(default=0, description="Unique URLs (deduplicated)")
    crawl_depth_reached: int = Field(default=0, description="Maximum crawl depth reached")
    pages_visited: int = Field(default=0, description="Total pages visited during crawl")
    forms_discovered: int = Field(default=0, description="Total forms found")
    api_endpoints: Set[str] = Field(default_factory=set, description="Potential API endpoints discovered")
    javascript_urls: int = Field(default=0, description="URLs discovered via JavaScript analysis")

    class Config:
        arbitrary_types_allowed = True
        json_encoders = {
            set: list,  # Convert sets to lists for JSON serialization
            datetime: lambda v: v.isoformat()
        }

    def add_url(self, url: DiscoveredURL):
        """Add a discovered URL to results"""
        self.urls.append(url)
        self.total_urls += 1

    def add_form(self, form: FormData):
        """Add a discovered form to results"""
        self.forms.append(form)
        self.forms_discovered += 1

    def get_unique_paths(self) -> Set[str]:
        """Extract unique URL paths from discovered URLs"""
        from urllib.parse import urlparse
        return {urlparse(url.url).path for url in self.urls}

    def get_urls_by_depth(self, depth: int) -> List[DiscoveredURL]:
        """Get URLs discovered at specific depth"""
        return [url for url in self.urls if url.depth == depth]

    def get_post_endpoints(self) -> List[DiscoveredURL]:
        """Get all POST endpoints discovered"""
        return [url for url in self.urls if url.method == "POST"]

    def get_forms_by_method(self, method: str) -> List[FormData]:
        """Get forms filtered by HTTP method"""
        return [form for form in self.forms if form.method.upper() == method.upper()]

File: src/models/url.ts
/**
 * URL discovery data models for deep crawling with Playwright
 */

import { z } from 'zod';

/**
 * Individual discovered URL from deep crawling
 */
export const discoveredURLSchema = z.object({
  /** Full URL */
  url: z.string().url(),
  /** HTTP method (GET, POST, etc.) */
  method: z.string().default('GET'),
  /** Page where URL was found */
  sourcePage: z.string().url().optional(),
  /** Crawl depth from start */
  depth: z.number().int().nonnegative().default(0),
  /** Query params or form fields */
  parameters: z.record(z.any()).default({}),
  /** Timestamp when URL was discovered */
  discoveredAt: z.date().default(() => new Date()),
  /** HTTP response code */
  responseCode: z.number().int().min(100).max(599).optional(),
  /** Content-Type header */
  contentType: z.string().optional(),
  /** Discovery method: deep_crawl|form|js_analysis */
  discoveredVia: z.string().default('deep_crawl'),
  /** Discovery source (alias for discoveredVia) */
  source: z.string().optional(),
});

export type DiscoveredURL = z.infer<typeof discoveredURLSchema>;

/**
 * Form metadata discovered during crawling
 */
export const formDataSchema = z.object({
  /** Form action URL */
  actionUrl: z.string().url(),
  /** Form submission method */
  method: z.string().default('GET'),
  /** Form fields with metadata */
  fields: z.array(z.record(z.any())).default([]),
  /** Page containing the form */
  sourcePage: z.string().url().optional(),
  /** Whether form requires authentication */
  requiresAuth: z.boolean().optional(),
});

export type FormData = z.infer<typeof formDataSchema>;

/**
 * Results from URL discovery crawl using Playwright
 */
export const urlDiscoveryResultSchema = z.object({
  /** All discovered URLs */
  urls: z.array(discoveredURLSchema).default([]),
  /** Discovered forms */
  forms: z.array(formDataSchema).default([]),
  /** Total URLs discovered */
  totalUrls: z.number().int().nonnegative().default(0),
  /** Unique URLs (deduplicated) */
  uniqueUrls: z.number().int().nonnegative().default(0),
  /** Maximum crawl depth reached */
  crawlDepthReached: z.number().int().nonnegative().default(0),
  /** Crawl depth (alias for crawlDepthReached) */
  crawlDepth: z.number().int().nonnegative().optional(),
  /** Total pages visited during crawl */
  pagesVisited: z.number().int().nonnegative().default(0),
  /** Total forms found */
  formsDiscovered: z.number().int().nonnegative().default(0),
  /** Potential API endpoints discovered */
  apiEndpoints: z.array(z.string()).default([]),
  /** URLs discovered via JavaScript analysis */
  javascriptUrls: z.number().int().nonnegative().default(0),
  /** Target URL for the crawl */
  targetUrl: z.string().optional(),
});

export type URLDiscoveryResult = z.infer<typeof urlDiscoveryResultSchema>;

/**
 * Helper functions for URL discovery results
 */
export class URLDiscoveryResultHelper {
  /**
   * Add a discovered URL to results
   */
  static addUrl(result: URLDiscoveryResult, url: DiscoveredURL): URLDiscoveryResult {
    return {
      ...result,
      urls: [...result.urls, url],
      totalUrls: result.totalUrls + 1,
    };
  }

  /**
   * Add a discovered form to results
   */
  static addForm(result: URLDiscoveryResult, form: FormData): URLDiscoveryResult {
    return {
      ...result,
      forms: [...result.forms, form],
      formsDiscovered: result.formsDiscovered + 1,
    };
  }

  /**
   * Extract unique URL paths from discovered URLs
   */
  static getUniquePaths(result: URLDiscoveryResult): Set<string> {
    return new Set(result.urls.map((url) => new URL(url.url).pathname));
  }

  /**
   * Get URLs discovered at specific depth
   */
  static getUrlsByDepth(result: URLDiscoveryResult, depth: number): DiscoveredURL[] {
    return result.urls.filter((url) => url.depth === depth);
  }

  /**
   * Get all POST endpoints discovered
   */
  static getPostEndpoints(result: URLDiscoveryResult): DiscoveredURL[] {
    return result.urls.filter((url) => url.method === 'POST');
  }

  /**
   * Get forms filtered by HTTP method
   */
  static getFormsByMethod(result: URLDiscoveryResult, method: string): FormData[] {
    return result.forms.filter((form) => form.method.toUpperCase() === method.toUpperCase());
  }

  /**
   * Get unique endpoints (deduplicated URLs)
   */
  static getUniqueEndpoints(result: URLDiscoveryResult): DiscoveredURL[] {
    const seen = new Set<string>();
    return result.urls.filter((url) => {
      if (seen.has(url.url)) {
        return false;
      }
      seen.add(url.url);
      return true;
    });
  }
}
